File: ops/src/infinicore/ops/causal_softmax/causal_softmax.cc
#include "infinicore/ops/causal_softmax.hpp"
#include "../../utils.hpp"

namespace infinicore::op {

INFINICORE_GRAPH_OP_DISPATCHERS_IMPL(CausalSoftmax);

CausalSoftmax::CausalSoftmax(Tensor output, const Tensor &input) {
    INFINICORE_ASSERT_TENSORS_SAME_DEVICE(output, input);
    INFINICORE_GRAPH_OP_DISPATCH(output->device().getType(), output, input);
}

void CausalSoftmax::execute(Tensor output, const Tensor &input) {
    INFINICORE_GRAPH_OP_RECORD_OR_RUN(CausalSoftmax, output, input);
}

Tensor causal_softmax(const Tensor &input) {
    auto output = Tensor::empty(input->shape(), input->dtype(), input->device());
    causal_softmax_(output, input);
    return output;
}

void causal_softmax_(Tensor output, const Tensor &input) {
    CausalSoftmax::execute(output, input);
}

} // namespace infinicore::op

File: python/infinicore/nn/functional/causal_softmax.py
from infinicore.lib import _infinicore
from infinicore.tensor import Tensor


def causal_softmax(input: Tensor, out=None) -> Tensor:
    r"""Apply a causal softmax function."""

    if out is None:
        return Tensor(_infinicore.causal_softmax(input._underlying))

    _infinicore.causal_softmax_(out._underlying, input._underlying)

    return out
